controllers/dataplane_controller.go

```diff
@@ -0,0 +1,154 @@
+package controllers
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	dataplane "github.com/kong/blixt/internal/dataplane/client"
+	"github.com/kong/blixt/pkg/vars"
+)
+
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways/finalizers,verbs=update
+
+//+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=services/status,verbs=get
+
+//+kubebuilder:rbac:groups=core,resources=endpoints,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=endpoints/status,verbs=get
+
+// DataplaneReconciler reconciles the dataplane pods.
+type DataplaneReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	BackendsClientManager *dataplane.BackendsClientManager
+
+	updates chan event.GenericEvent
+}
+
+func NewDataplaneReconciler(client client.Client, schema *runtime.Scheme, manager *dataplane.BackendsClientManager) *DataplaneReconciler {
+	return &DataplaneReconciler{
+		Client:                client,
+		Scheme:                schema,
+		BackendsClientManager: manager,
+		updates:               make(chan event.GenericEvent, 1000),
```

Comment:
Yeah I think this should work :thinking: 

At this point in the code if we are guaranteed to have the backends updated in the client, then it seems to follow that we should be able to "collapse" stacked up events by simply dropping extras, since all messages mean the same thing and trigger the same update. Let's give that a try :+1:

Reply:
Awesome!